### DISCUSSION TITLE
Modifying Cell Aspect Ratio

### DISCUSSION THREAD

**Discussion by JhonCordova** [AMReX-Combustion/PeleC#853] [Q&A] [answered]
Hi, Everyone

I noticed that the cell aspect ratio is always set to 1 (ΔX=ΔY=ΔZ​). Is there any way to adjust this so that ΔX≠ΔY≠ΔZ​? If not, how hard would it be to implement such a feature?

Thanks in advance!

**Comment by baperry2** [comment]
Right now, ΔX=ΔY=ΔZ​ is a firm requirement in PeleC. 

It would definitely be useful to relax this for many problems. That's an idea we've thrown around but I'm not aware of anyone having gone through thoroughly to figure out what would actually be required. There are definitely some places in the code that are already set up to allow differing grid spacing for each direction, but there probably are other places where the shortcut of treating them as equal has been used. I'm not sure if allowing differing grid spacing would merely be an issue of going through and ensuring proper bookkeeping is done everywhere for ΔX≠ΔY≠ΔZ​ (probably not trivial to do, but feasible), or if there would be more major roadblocks anywhere (e.g. using capability from AMReX that would not support this). 

@marchdf do you remember why the limitation of ΔX=ΔY=ΔZ​ was originally enforced? Do you know of any major roadblocks that would prevent this from being relaxed?

**Reply by marchdf** [reply]
I think it was a thing to do with EB. I wouldn't be totally surprised if non-EB mostly just worked without that. There's probably just a couple of places where that assumption is explicitly made.

**Reply by baperry2** [reply]
Yeah, it having something to do with EB rings a bell for me as well. 

@JhonCordova, I'd recommend starting with a non-EB case, remove the error for ΔX≠ΔY≠ΔZ​, and then try a case with non-unity aspect ratio and see what (if anything) breaks. You can turn off different types of physics using these flags to see what works and what doesn't:
```
pelec.do_hydro = 1
pelec.do_react = 1
pelec.diffuse_temp=1
pelec.diffuse_enth=1
pelec.diffuse_spec=1
pelec.diffuse_vel=1
```

**Reply by drummerdoc** [reply, marked as answer]
It would be a bug if it required equal cell sizes in all direction without EB.  I fully expect that it should work.  We should modify the assertion to enforce this only if EB is being actually used.....my $.02

**Reply by JhonCordova** [reply]
Thanks. It is working

**Reply by drummerdoc** [reply]
Excellent.  Is there any test changes we should push to formalize this?  Thanks for checking back @JhonCordova

**Reply by JhonCordova** [reply]
The simulation is running for the channel flow case, but I still need to include some modifications, such as a variable external force to maintain the Reynolds number. As soon as I obtain results that can be compared with DNS data, I will let you know.

**Reply by baperry2** [reply]
Still a bit worried that there's possibly something buried where we don't account for different dx in different directions. That would likely cause a silent issue where the simulation runs to completion but potentially gives a bad result. So the validation comparisons will be important.

**Reply by JhonCordova** [reply]
Hi everyone,

I wanted to share a quick update on the channel flow case I’ve been working on.

The simulation now runs correctly with non-cubic cells (ΔX ≠ ΔY ≠ ΔZ) and no EB, as previously discussed. However, I’ve run into a few issues:

1. External force implementation (prob.cpp):
    I’ve had some difficulty tuning the external forcing to maintain the target Reynolds number. I'm currently using a time-varying force, but I’m not sure it’s optimal. If anyone has suggestions or example setups, I’d really appreciate them.

2. Validation results:
    Without LES, the simulation produces results that qualitatively agree with [DNS ](https://www.researchgate.net/publication/231941657_Direct_numerical_simulation_of_compressible_turbulent_channel_flow_between_adiabatic_and_isothermal_walls)— velocity profiles and general trends are similar — but there are still some deviations. I’m currently checking if these are related to resolution, forcing, or boundary conditions.

![Figure1](https://github.com/user-attachments/assets/15442be4-73a5-4f26-8a12-9675a20bcac0)
![figure2](https://github.com/user-attachments/assets/fd1bac21-440d-4e59-ab3c-7e874edb6e3b)
![Figure3](https://github.com/user-attachments/assets/46e06c35-d86f-4ad7-abc8-ebe7e4630dec)

3. LES issue with non-cubic cells:
    When I enable LES (Smagorinsky model), the simulation runs, but the temperature drops to unphysically low values (~50 K). I’ve confirmed that this doesn’t happen in the non-LES runs. Interestingly, the issue disappears when I switch to cubic cells (ΔX = ΔY = ΔZ), which strongly suggests that something in the LES model or its interaction with the grid spacing is not correctly handling anisotropic grids. I’ve verified that the implementation of the model is consistent with [previous formulations](https://www.researchgate.net/publication/225746215_Subgrid-Scale_Models_for_Compressible_Large-Eddy_Simulations), and I also tried defining the filter width as the cubic root of cell volume, but the problem persists unless cubic cells are used.

If anyone has seen similar behavior or can point me to parts of the code where Δ is assumed isotropic under LES, I’d be very grateful. I’ll keep you posted as I continue investigating.

Here are the case files:

[CMakeLists.txt](https://github.com/user-attachments/files/20211796/CMakeLists.txt)
[GNUmakefile.txt](https://github.com/user-attachments/files/20211808/GNUmakefile.txt)
[inputs.txt](https://github.com/user-attachments/files/20211830/inputs.txt)
[Makepackage.txt](https://github.com/user-attachments/files/20211864/Makepackage.txt)
[probcpp.txt](https://github.com/user-attachments/files/20211885/probcpp.txt)
[probH.txt](https://github.com/user-attachments/files/20211886/probH.txt)
[prob_parmH.txt](https://github.com/user-attachments/files/20211887/prob_parmH.txt)

Thanks again for all your support.

**Reply by baperry2** [reply]
Nice work! Good to know that things mostly appear to be working with `dx != dy != dz`. 

Responses to your questions:
1) For forcing, I suppose what you want is for the forcing to exactly cancel the integrated wall shear stress. You'd have to hack a bit, but you can probably compute the integrated wall stress after it is computed in Diffusion.cpp and then use that to set an exactly balancing external forcing across the entire domain, keeping the total momentum and therefore Re constant. 
2) Hopefully this is resolved either by adding resolution or adjusting the forcing
3) I was going to suggest that it might be in setting the lengthscale. I don't have a good guess of where it could be, but it wouldn't surprise me if there is somewhere that has implicitly assumed `dx = dy = dz` within the LES code because it hasn't ever been tested on anisotropic grids.

**Reply by yjh124** [reply]
<img width="323" height="75" alt="image" src="https://github.com/user-attachments/assets/e610f51c-89e5-4b8d-90ff-5bfc442f77ed" />
<img width="630" height="160" alt="image" src="https://github.com/user-attachments/assets/d55639c1-9918-4e98-bfe9-5a5da3c9e2ba" />

**Reply by yjh124** [reply]
I know why the friction coefficient and friction Reynolds number are lower here—it's because the physical property model is incorrect. In the literature, Cp and Cv are constants, but in the Fuego model, Cp and Cv are functions of temperature T.

**Reply by JhonCordova** [reply]
In PeleC, EOS/transport pairings are strictly constrained by the code. In particular, a constant-property setup with Sutherland viscosity is not supported:

- GammaLaw supports Constant transport only.
- Fuego and Soave–Redlich–Kwong (SRK) are intended to be used with Simple transport.
- Combinations such as GammaLaw/Sutherland or GammaLaw/Simple are explicitly disallowed and fail at compile time.

As a result, reproducing the constant-Cp and constant-Cv channel-flow setups commonly used in the literature is currently not possible without modifying the EOS/transport infrastructure.

**Reply by yjh124** [reply]
Perhaps we can change the viscosity calculation in constant.H to use the Sutherland formula, adopting the combination of gammalaw + null + constant.

**Comment by yjh124** [comment]
Hello, I wonder if the problem has been solved? How was it solved? I have encountered the same problem. Thank you very much.

**Reply by baperry2** [reply]
The isotropic grid size restriction has been lifted for non-EB simulations (#964) but remains in place for EB simulations despite some progress in #967. Further development in that direction is not planned but contributions would be welcome.

**Reply by yjh124** [reply]
Thanks very much for the informative reply and for sharing the PR links. I'll note the current status and look into it further.
